Component System: Upgrade FontSizePicker

ItsJonQ opened the pull request:
This update adapts `FontSizePicker` from `@wordpress/components` with the new Component System setup.
It does this by adding the (current) external `@wp-g2` component system packages as dependencies and using them directly.

This strategy was something @gziolo had suggested here:
https://github.com/WordPress/gutenberg/issues/27484#issuecomment-738857333


## How has this been tested?

* `npm install`
* Run `npm run dev`
* Fire up local Gutenberg
* Edit or add a Paragraph block
* Adjust the font-size
* It should work like normal (and the controls may look slightly slightly different)

## Screenshots

![wp-g2-font-size-picker](https://user-images.githubusercontent.com/2322354/101541946-0ea8dc80-3970-11eb-8259-59f55e22d752.gif)


## Types of changes

* Adding `@wp-g2` dependencies
* Add `COMPONENT_SYSTEM_PHASE` env flag
* Created a context/adapter
* Connected `FontSizePicker` to the component system context

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

aristath commented:
Since we're doing this, would it be possible to account for other units as well and not limit this to just pixels? :heart:

ItsJonQ commented:
> would it be possible to account for other units as well and not limit this to just pixels?

Indeed! The underlying `UnitInput` supports it. The missing piece would be... 

* figuring out the control design to accommodate unit values (e.g. making sure the input is wide enough)
* ensuring the value flows correctly from the hooks/styles data into the control

jorgefilipecosta commented on packages/components/src/font-size-picker/next.js:
```diff
@@ -0,0 +1,13 @@
+import { __unstableWithNext as withNext } from '../__next/context';
+
+export function withNextComponent( current ) {
+	return withNext( current, FontSizeControl, 'WPComponentsFontSizePicker' );
```
I wonder if for something like a FontSizePicker we really need to keep two versions. I guess in this case would be ok if we just deleted the current component and used the new one?

ItsJonQ replied:
That's up to us really. It looks like `FontSizePicker` is only really used in one spot, that being the `hooks` typography tools.

The thing that is uncertain is if it's used by 3rd party blocks. In that case, I feel like it may be safer to...

1. Start with having both
2. Ensure it's stable
3. Create some documentation for 3rd party devs regarding this migration (context/adapter). Announce this.
4. See if there's any pushback
5. If there are many people who want to keep the old `FontSizePicker`, provide 3rd party devs with the ability to "opt-out" (with maybe deprecation notices)
6. Wait a bit
7. Remove old implementation

☝️ Repeat for components as needed

gziolo commented on webpack.config.js:
```diff
@@ -97,6 +97,13 @@ module.exports = {
+				parseInt(
+					process.env.npm_package_config_COMPONENT_SYSTEM_PHASE,
+					10
+				) || 1
```
Noting that `COMPONENT_SYSTEM_PHASE` is enabled at the moment.

When I disable it I see the following error when trying to insert the Paragraph block:

![Uploading Screen Shot 2020-12-09 at 21.41.48.png…]()

gziolo replied:
I can confirm it's resolved now 👍🏻

gziolo commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -52,14 +53,17 @@ function getSelectOptions( optionsArray, disableCustomFontSizes ) {
+		value,
+		withSlider = false,
+	},
+	ref
```
it's probably the place where the error I shared is triggered. Maybe it requires a fallback to `createRef()`?

gziolo commented on packages/components/src/__next/context/component-system-provider.js:
```diff
@@ -0,0 +1,23 @@
+	children,
+	value = {},
+} ) {
+	const contextValue = { ...value };
```
It needs to be wrapped with:

```if ( process.env.COMPONENT_SYSTEM_PHASE === 1 ) {```

gziolo commented on packages/components/src/font-size-picker/next.js:
```diff
@@ -0,0 +1,13 @@
+import { __unstableWithNext as withNext } from '../__next/context';
+
+export function withNextComponent( current ) {
+	return withNext( current, FontSizeControl, 'WPComponentsFontSizePicker' );
```
It needs to be wrapped with:

`if ( process.env.COMPONENT_SYSTEM_PHASE === 1 ) {`

An alternative I tried is to pass a string name and keep the mapping inside `withNext`. It works but then you need to usually update code in two places.

jorgefilipecosta reviewed (commented):
Hi @ItsJonQ nice work on this PR.
It seems we have a good approach to enhance our components 👍 
I think it would be nice to move the font size picker to the Gutenberg repository, e.g: if someone needs to debug the code to fix a bug the code of the component is not in this repository.

The component would then use other g2 components inside.

The prototype of g2 includes support for other units and even complex css calculations would it be possible to include the support for this or you prefer to do that in a follow-up PR?

gziolo commented on packages/components/package.json:
```diff
@@ -45,6 +45,10 @@
 		"@wordpress/primitives": "file:../primitives",
 		"@wordpress/rich-text": "file:../rich-text",
 		"@wordpress/warning": "file:../warning",
+		"@wp-g2/components": "^0.0.113",
+		"@wp-g2/styles": "^0.0.113",
+		"@wp-g2/substate": "^0.0.113",
+		"@wp-g2/utils": "^0.0.113",
```
We need only the following so far:
```suggestion
		"@wp-g2/components": "^0.0.113",
```

gziolo commented on packages/block-editor/src/hooks/font-size.js:
```diff
@@ -129,9 +120,20 @@ export function FontSizeEdit( props ) {
+	const fontSizeValue =
+		fontSizeObject?.size || style?.typography?.fontSize || fontSize;
+
+	return <FontSizePicker onChange={ onChange } value={ fontSizeValue } />;
```
Is this mapping necessary here? Wouldn't it be something to put inside the adapter function?

ItsJonQ replied:
> Is this mapping necessary here?

@gziolo I believe so. `style.typography.fontSize` is very specific to the `hooks/` part of the editor. After testing, it appears that `fontSizeValue` my change depending on values defined in the `theme.json` setup. If that's correct,  the prop "remap" feels correct in this `hooks/` area (vs. in the components package).

gziolo commented:
I already see some good impact with the latest changes applied to `@wp-g2/components`. It's currently at
> Size Change: +104 kB (7%) 🔍
for `@wordpress/components` entry point in Gutenberg. It was 112 kB yesterday 😄

gziolo commented on package-lock.json:
```diff
@@ -14239,6 +14245,247 @@
+						"supports-color": "^5.3.0"
+					}
+				},
+				"commander": {
```
It's surprising to see `commander`, `postcss` and `rtlcss` on the list of dependencies. Is any of them used on the front-end?

gziolo commented:
This error is strange:

<img width="858" alt="Screen Shot 2020-12-10 at 07 40 51" src="https://user-images.githubusercontent.com/699132/101730914-25a81580-3abb-11eb-9dba-9b0010361782.png">

It's possible that one of the npm packages bundled with G2 components should be transpiled with Babel.

gziolo commented:
With 82535b6 we are now at and I only aligned versions for `react-use-gesture`:
> Size Change: +99.5 kB (7%) 🔍

@ItsJonQ, we should upgrade all libraries that Gutenberg and G2 components share to the latest versions in Gutenberg in a separate PR. Do the same for G2 components. Regenerate the lock file and see how it looks. Maybe we can reduce the size of new code added to a reasonable size and do a more aggressive rollout :)

youknowriad commented on package.json:
```diff
@@ -19,7 +19,8 @@
 	"config": {
-		"GUTENBERG_PHASE": 2
+		"GUTENBERG_PHASE": 2,
+		"COMPONENT_SYSTEM_PHASE": 0
```
Why is this flag used for? In't it redundunt with the "context" component provider?

gziolo replied:
We want to merge this PR with all the new code removed (100kb as of today according to the CI job) when running `npm run build`. This way we can integrate a few components, remove code duplication wherever possible and enable this flag (or remove completely) afterwards.

youknowriad replied:
What's the point of merging components and not really use them in the plugin?

gziolo replied:
I can also confirm now that Tree Shaking works like a charm:
<img width="620" alt="Screen Shot 2020-12-10 at 09 18 48" src="https://user-images.githubusercontent.com/699132/101740332-f8fafa80-3ac8-11eb-8f67-0959b6c86d05.png">

🎉

gziolo commented:
I'm happy to report that Tree Shaking is working perfectly fine 🎉 

<img width="620" alt="Screen Shot 2020-12-10 at 09 18 48" src="https://user-images.githubusercontent.com/699132/101740332-f8fafa80-3ac8-11eb-8f67-0959b6c86d05.png">

🎉 

Now, I will enable the feature flag again. Locally, I see the following error with the flag enabled:

<img width="1694" alt="Screen Shot 2020-12-10 at 08 22 58" src="https://user-images.githubusercontent.com/699132/101741072-e7662280-3ac9-11eb-87a3-c809d9e3063c.png">

Let's see if it's confirmed on Travis. I was able to reproduce the same issue without my latest commits.

gziolo commented:
We are back at:

<img width="633" alt="Screen Shot 2020-12-10 at 10 18 17" src="https://user-images.githubusercontent.com/699132/101751832-0b793200-3ad1-11eb-9f5e-b1351bccf3f5.png">

Some e2e tests are failing so it must be something with the recent changes to the G2 integration.

gziolo commented:
Each individual block's CSS has some duplicated CSS:

### Audio
```css
:root{--wp-admin-theme-color:#007cba;--wp-admin-theme-color-darker-10:#006ba1;--wp-admin-theme-color-darker-20:#005a87}.wp-block-audio figcaption{margin-top:.5em;margin-bottom:1em}.wp-block-audio audio{width:100%;min-width:300px}
```

### Spacer
```css
:root{--wp-admin-theme-color:#007cba;--wp-admin-theme-color-darker-10:#006ba1;--wp-admin-theme-color-darker-20:#005a87}.wp-block-spacer{clear:both}
```

### Verse
```css
:root{--wp-admin-theme-color:#007cba;--wp-admin-theme-color-darker-10:#006ba1;--wp-admin-theme-color-darker-20:#005a87}pre.wp-block-verse{font-family:inherit}
```

It's also in `common.css`:
```css
:root{--wp-admin-theme-color:#007cba;--wp-admin-theme-color-darker-10:#006ba1;--wp-admin-theme-color-darker-20:#005a87}:root .has-pale-pink-background-color{background-color:#f78da7}:root .has-vivid-red-background-color{background-color:#cf2e2e}:root .has-luminous-vivid-orange-background-color{background-color:#ff6900}:root .has-luminous-vivid-amber-background-color{background-color:#fcb900}:root .has-light-green-cyan-background-color{background-color:#7bdcb5}:root .has-vivid-green-cyan-background-color{background-color:#00d084}:root .has-pale-cyan-blue-background-color{background-color:#8ed1fc}:root .has-vivid-cyan-blue-background-color{background-color:#0693e3}:root .has-vivid-purple-background-color{background-color:#9b51e0}:root .has-white-background-color{background-color:#fff}:root .has-very-light-gray-background-color{background-color:#eee}:root .has-cyan-bluish-gray-background-color{background-color:#abb8c3}:root .has-very-dark-gray-background-color{background-color:#313131}:root .has-black-background-color{background-color:#000}:root .has-pale-pink-color{color:#f78da7}:root .has-vivid-red-color{color:#cf2e2e}:root .has-luminous-vivid-orange-color{color:#ff6900}:root .has-luminous-vivid-amber-color{color:#fcb900}:root .has-light-green-cyan-color{color:#7bdcb5}:root .has-vivid-green-cyan-color{color:#00d084}:root .has-pale-cyan-blue-color{color:#8ed1fc}:root .has-vivid-cyan-blue-color{color:#0693e3}:root .has-vivid-purple-color{color:#9b51e0}:root .has-white-color{color:#fff}:root .has-very-light-gray-color{color:#eee}:root .has-cyan-bluish-gray-color{color:#abb8c3}:root .has-very-dark-gray-color{color:#313131}:root .has-black-color{color:#000}:root .has-vivid-cyan-blue-to-vivid-purple-gradient-background{background:linear-gradient(135deg,#0693e3,#9b51e0)}:root .has-vivid-green-cyan-to-vivid-cyan-blue-gradient-background{background:linear-gradient(135deg,#00d084,#0693e3)}:root .has-light-green-cyan-to-vivid-green-cyan-gradient-background{background:linear-gradient(135deg,#7adcb4,#00d082)}:root .has-luminous-vivid-amber-to-luminous-vivid-orange-gradient-background{background:linear-gradient(135deg,#fcb900,#ff6900)}:root .has-luminous-vivid-orange-to-vivid-red-gradient-background{background:linear-gradient(135deg,#ff6900,#cf2e2e)}:root .has-very-light-gray-to-cyan-bluish-gray-gradient-background{background:linear-gradient(135deg,#eee,#a9b8c3)}:root .has-cool-to-warm-spectrum-gradient-background{background:linear-gradient(135deg,#4aeadc,#9778d1 20%,#cf2aba 40%,#ee2c82 60%,#fb6962 80%,#fef84c)}:root .has-blush-light-purple-gradient-background{background:linear-gradient(135deg,#ffceec,#9896f0)}:root .has-blush-bordeaux-gradient-background{background:linear-gradient(135deg,#fecda5,#fe2d2d 50%,#6b003e)}:root .has-purple-crush-gradient-background{background:linear-gradient(135deg,#34e2e4,#4721fb 50%,#ab1dfe)}:root .has-luminous-dusk-gradient-background{background:linear-gradient(135deg,#ffcb70,#c751c0 50%,#4158d0)}:root .has-hazy-dawn-gradient-background{background:linear-gradient(135deg,#faaca8,#dad0ec)}:root .has-pale-ocean-gradient-background{background:linear-gradient(135deg,#fff5cb,#b6e3d4 50%,#33a7b5)}:root .has-electric-grass-gradient-background{background:linear-gradient(135deg,#caf880,#71ce7e)}:root .has-subdued-olive-gradient-background{background:linear-gradient(135deg,#fafae1,#67a671)}:root .has-atomic-cream-gradient-background{background:linear-gradient(135deg,#fdd79a,#004a59)}:root .has-nightshade-gradient-background{background:linear-gradient(135deg,#330968,#31cdcf)}:root .has-midnight-gradient-background{background:linear-gradient(135deg,#020381,#2874fc)}:root .has-link-color a{color:#00e;color:var(--wp--style--color--link,#00e)}.has-small-font-size{font-size:.8125em}.has-normal-font-size,.has-regular-font-size{font-size:1em}.has-medium-font-size{font-size:1.25em}.has-large-font-size{font-size:2.25em}.has-huge-font-size,.has-larger-font-size{font-size:2.625em}.has-text-align-center{text-align:center}.has-text-align-left{text-align:left}.has-text-align-right{text-align:right}#end-resizable-editor-section{display:none}.aligncenter{clear:both}
```

and also `theme.css`:
```css
:root{--wp-admin-theme-color:#007cba;--wp-admin-theme-color-darker-10:#006ba1;--wp-admin-theme-color-darker-20:#005a87}#start-resizable-editor-section{display:none}.wp-block-audio figcaption{color:#555;font-size:13px;text-align:center}.is-dark-theme .wp-block-audio figcaption{color:hsla(0,0%,100%,.65)}.wp-block-code{font-family:Menlo,Consolas,monaco,monospace;color:#1e1e1e;padding:.8em 1em;border:1px solid #ddd;border-radius:4px}.wp-block-embed figcaption{color:#555;font-size:13px;text-align:center}.is-dark-theme .wp-block-embed figcaption{color:hsla(0,0%,100%,.65)}.blocks-gallery-caption{color:#555;font-size:13px;text-align:center}.is-dark-theme .blocks-gallery-caption{color:hsla(0,0%,100%,.65)}.wp-block-image figcaption{color:#555;font-size:13px;text-align:center}.is-dark-theme .wp-block-image figcaption{color:hsla(0,0%,100%,.65)}.wp-block-pullquote{border-top:4px solid;border-bottom:4px solid;margin-bottom:1.75em;color:currentColor}.wp-block-pullquote__citation,.wp-block-pullquote cite,.wp-block-pullquote footer{color:currentColor;text-transform:uppercase;font-size:.8125em;font-style:normal}.wp-block-navigation ul,.wp-block-navigation ul li{list-style:none}.wp-block-navigation-link.wp-block-navigation-link{margin:0}.wp-block-quote{border-left:.25em solid;margin:0 0 1.75em;padding-left:1em}.wp-block-quote__citation,.wp-block-quote cite,.wp-block-quote footer{color:currentColor;font-size:.8125em;margin-top:1em;position:relative;font-style:normal}.wp-block-quote.has-text-align-right{border-left:none;border-right:.25em solid;padding-left:0;padding-right:1em}.wp-block-quote.has-text-align-center{border:none;padding-left:0}.wp-block-quote.is-large,.wp-block-quote.is-style-large{border:none}.wp-block-search .wp-block-search__label{font-weight:700}.wp-block-group.has-background{padding:20px 30px;margin-top:0;margin-bottom:0}.wp-block-separator{border:none;border-bottom:2px solid;margin-left:auto;margin-right:auto;opacity:.4}.wp-block-separator:not(.is-style-wide):not(.is-style-dots){max-width:100px}.wp-block-separator.has-background:not(.is-style-dots){border-bottom:none;height:1px}.wp-block-separator.has-background:not(.is-style-wide):not(.is-style-dots){height:2px}.wp-block-table{border-collapse:collapse}.wp-block-table thead{border-bottom:3px solid}.wp-block-table tfoot{border-top:3px solid}.wp-block-table td,.wp-block-table th{padding:.5em;border:1px solid;word-break:normal}.wp-block-table figcaption{color:#555;font-size:13px;text-align:center}.is-dark-theme .wp-block-table figcaption{color:hsla(0,0%,100%,.65)}.wp-block-video figcaption{color:#555;font-size:13px;text-align:center}.is-dark-theme .wp-block-video figcaption{color:hsla(0,0%,100%,.65)}.wp-block-template-part.has-background{padding:20px 30px;margin-top:0;margin-bottom:0}#end-resizable-editor-section{display:none}
```

Should individual block's CSS files depend on the common file?

ItsJonQ commented:
Took a look at the build file. From what I can tell, the `const` is coming from [zustand](https://github.com/pmndrs/zustand) (a dependency of `@wp-g2/substate`).

<img width="1356" alt="Screen Shot 2020-12-10 at 11 56 36 AM" src="https://user-images.githubusercontent.com/2322354/101803677-c65bfc80-3ade-11eb-9d90-6d6c3db8cc05.png">

It's really weird... because `zustand` has a `.cjs.js` export:

```
  "main": "index.cjs.js",
```

That one doesn't have `const` in it.

🤔 

Not sure how to fix this one.

gziolo commented:
I think webpack defaults to `module` field rather than `main` with our setup. You probably can add an exception to include this package in Babel transpilation.

ItsJonQ commented:
@gziolo Interesting!

> You probably can add an exception to include this package in Babel transpilation.

How do I do that? 🙈 😂

youknowriad commented:
or replace it with useReducer or something like that. I believe state libraries are needed when you have a big application with a lot of nesting. For components, I'd personally avoid centralized state and only rely on local state.

ItsJonQ commented:
> For components, I'd personally avoid centralized state and only rely on local state.

The components in `@wp-g2/components` are using localized state (not globalized store). The `@wp-g2/substate` package optimizes `zustand` for that purpose :).

ItsJonQ commented:
E2E tests would be interesting.. as the component coming from `@wp-g2` won't have the same markup structure or className structure.

So tests looking for selectors like `.components-font-size-picker__controls .components-font-size-picker__number` will fail.

We'll have to update them.

The components from this system automatically render namespaced HTML attributes like this:

```
data-g2-component="FormGroup"
```

There's system supported ways for setting/getting this specific HTML attribute (without having to manually write it).
This is so that we may update the underlying `html-attribute-name` without having to do a mass find/replace.

If targeting things in this way sounds good, I can maybe create some helper functions for Puppeteer to be able to query these component selectors using this structure.

ItsJonQ commented on packages/components/src/font-size-picker/next.js:
```diff
@@ -0,0 +1,13 @@
+/**
+ * External dependencies
+ */
+import { FontSizeControl } from '@wp-g2/components';
```
> Could we bring the code of FontSizeControl to Gutenberg

To do that, we would need (at minimum) all of the Components (and systems) that make up that control.
I've included a screenshot below of a high-level overview. Required components are in the yellow squares:

<img width="1639" alt="Screen Shot 2020-12-10 at 2 51 27 PM" src="https://user-images.githubusercontent.com/2322354/101822362-34acb900-3af7-11eb-9ee0-b2e88b43d056.png">

Link to Miro board (source of screenshot)
https://miro.com/app/board/o9J_khVkF9A=/?moveToWidget=3074457351837142359&cot=14

---

Adding the systems and components to build up to FontSizeControl was the original proposal.
However, @gziolo suggested we could try consuming the `@wp-g2/components` package directly to expedite integration, which I think is a good idea (assuming important things like bundling aren't severely impacted)

youknowriad replied:
Can't we add FontSizeControl component (just this component)  and use its dependencies from the package? The idea is that this will allow us to review the component properly in isolation and step by step we'd do the same for all its dependencies.

jorgefilipecosta commented:
> E2E tests would be interesting.. as the component coming from `@wp-g2` won't have the same markup structure or className structure.
> 
> So tests looking for selectors like `.components-font-size-picker__controls .components-font-size-picker__number` will fail.
> 
> We'll have to update them.
> 
> The components from this system automatically render namespaced HTML attributes like this:
> 
> ```
> data-g2-component="FormGroup"
> ```
> 
> There's system supported ways for setting/getting this specific HTML attribute (without having to manually write it).
> This is so that we may update the underlying `html-attribute-name` without having to do a mass find/replace.
> 
> If targeting things in this way sounds good, I can maybe create some helper functions for Puppeteer to be able to query these component selectors using this structure.



I think relying on things like data-g2-component="FormGroup" to target makes our tests very dependent on the component system.
I think the tests should be updated to rely on things like aria-labels, Text of label elements and buttons, etc (probably relying more on XPath and less on CSS selectors). If people agree on this targeting approach I can propose a side PR that changes the tests failing here.

gziolo commented:
Regarding e2e test, I echo what @jorgefilipecosta said. There is a very recent issue created by @kevin940726 where it was discussed more in-depth: https://github.com/WordPress/gutenberg/issues/27260. There is also an older issue (closed) where the same topic was raised by @talldan: https://github.com/WordPress/gutenberg/issues/7974. The general sentiment is to avoid testing based on HTML parts that are implementation details. The other question is whether we really need a library yo help with that or could we introduce our own abstraction for missing features but that's what the first linked issue is trying to address.

ItsJonQ commented:
> I think the tests should be updated to rely on things like aria-labels, Text of label elements and buttons, etc

@gziolo + @jorgefilipecosta This I agree with, **especially for E2E** test :). I was offering the data attribute as a possible alternative.

(That being said, I think it's valuable to do deliberate HTML selector querying for unit tests at the Component System level, but that's another thing 😂 )

> The general sentiment is to avoid testing based on HTML parts that are implementation detail

In that case... how do we fix up the current tests?
I suppose we can refactor them to (try to) target elements based on labels/text? :)

saramarcondes commented:
Currently it's not possible to fix the E2E tests as the FontSizePicker component itself is broken when consumed by Gutenberg. The issue has to do with this line in G2: https://github.com/ItsJonQ/g2/blob/master/packages/create-styles/src/create-style-system/create-core-element.js#L141

We do automatic ref merging on that line. If `forwardedRef` is `undefined` (which it is in this case) then the `react-merge-refs` utility will fail. We should not pass undefined to `mergeRefs` anywhere. I'm actually unsure _why_ exactly we're doing ref merging in this case. `forwardedRef` is being treated as a prop, rather than an actual forwarded ref. The only places I know where we use `mergeRefs` is when we require a ref in the component itself and also want to continue forwarding refs (https://github.com/ItsJonQ/g2/blob/efa4bc41ba5e6a3169c3fe82fd7ee2b6381f3609/packages/components/src/TextInput/TextInput.js#L43 for example). `forwardedRef` is not a prop that is used anywhere inside G2.

Inside of `createCoreElement` however, I'm fairly certain it's not necessary to do this as there will only ever be the forwarded ref, which we'll simply pass to `React.createElement`.

I've opened a PR in G2 to fix this here: https://github.com/ItsJonQ/g2/pull/199. Once that PR is merged we can publish a new version of G2 and fix the error that @gziolo shared. Then we will be able to address the E2E issue (once the component actually renders).

jorgefilipecosta commented:
While trying to make the font size picker end to end test work with the version we have in master and this one I found some things in the markup we produce that I think we probably should fix:
![image](https://user-images.githubusercontent.com/11271197/102652379-bccd3700-4165-11eb-9ef1-3686668a873c.png)


After clicking to open the font size picker. This is the markup we see. There is an empty label, and that label contains a label for reference for an element whose id is not found.
I think we should not render that label at all, and the area should use a aria-labelledby to reference the "Font size" label that is visible (that's what happens on master).

I also noticed that in master each option is under a list while on g2 they are under a div. I guess in master we are using a list because it is more sematic. Should we also use a list on g2?

jorgefilipecosta commented:
The a11y I think we have (that also makes it harder to create end to end tests) is that just judging by the markup, I don't find any relation between the are that is opened to change the font size and the font size label/button that opens the area.
Another issue is that after we open the font size picker select a font size by moving the arrows but don't press enter and just tab to the next field the focus does not go to custom font size and is lost.

jorgefilipecosta commented:
Right now this branch is not working, e.g: when we add a paragraph we see a crash.
```
index.tsx:11 Uncaught TypeError: Cannot add property current, object is not extensible
    at index.tsx:11
    at Array.forEach (<anonymous>)
    at index.tsx:7
    at HTMLUnknownElement.callCallback (react-dom.82e849f1.js:182)
    at Object.invokeGuardedCallbackDev (react-dom.82e849f1.js:231)

```

Inside:
```
export default function mergeRefs<T = any>(
  refs: Array<React.MutableRefObject<T> | React.LegacyRef<T>>
): React.RefCallback<T> {
  return (value) => {
    refs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref != null) {
        (ref as React.MutableRefObject<T | null>).current = value;
      }
    });
  };
}
```


@saramarcondes could this error be related to https://github.com/ItsJonQ/g2/pull/199/ as it relates to merge ref?

saramarcondes commented:
@jorgefilipecosta Yup! As I described in https://github.com/WordPress/gutenberg/pull/27594#issuecomment-744662697 the issue is with that PR. I don't have time right now to spend on this issue but if someone re-publishes G2 with the new update I merged and then updates this PR it should fix the issue.

jorgefilipecosta committed:
Update min/max for FontSizeControl and improve getSelectOptions() to assign option object (+5 squashed commits)
Squashed commits:
[3b2e8cd71d] Update @wp-g2/components to latest v0.0.127
[144a98ed93] Add new font size picker to global styles
[80beed6b33] Remove additional props
[8d288e5177] Move font size component into gutenberg repository.
[1cf09cb4da] Component System: Upgrade FontSizePicker

gziolo commented:
@jorgefilipecosta, there is new version of G2 packages available.

youknowriad commented on packages/components/src/font-size-picker/next/font-size-control.js:
```diff
@@ -0,0 +1,79 @@
+		withNumberInput,
+		withSelect,
+		...otherProps
+	} = useFontSizeControl( props );
```
Can we document the props of the component. Are these the exact same props as the existing component? What are the differences. For me personally, this is the most important thing when adding the G2 components:  make sure the props make sense, that we don't have useless props and unnecessary public APIs. Also, ideally the props don't have any DOM related thing to make them cross devices.

jorgefilipecosta replied:
Hi @youknowriad, the component was adding some customization props but I removed them. Now the components just accept the same props as the existing one. The easiest way to see the props that are accepted is by checking the hook associated with the component.
https://github.com/WordPress/gutenberg/blob/86b3188dad0e454e8400dab2f08e1133047249e6/packages%2Fcomponents%2Fsrc%2Ffont-size-picker%2Fnext%2Fuse-font-size-control.js#L28

saramarcondes commented on packages/components/src/font-size-picker/next/use-font-size-control.js:
```diff
@@ -98,37 +83,20 @@ export function useFontSizeControl( props ) {
 
 	const isDefaultValue = ! is.defined( value );
 
-	const classes = cx( styles.FontSizeControl, className );
```
`className` is actually a requirement, it comes from `useContextSystem` and automatically applies the `wp-component-` and `component-` classes:

https://github.com/ItsJonQ/g2/blob/master/packages/context/src/use-context-system.js#L54

jorgefilipecosta replied:
Thank you @saramarcondes, I was not aware that className came from useContextSystem. I reverted the change 👍

jorgefilipecosta commented:
The end to end tests were updated to not rely on specific markup we had.

But the tests are still failing because of two real issues that exist:
- When we select the named font size, if we clean the custom input the font size does not resets.
	- In the debugging I did, the issue seems to be on the TextInput g2 component. It seems that if the value property changes by something outside the component, after that change the component does not fires onChange anymore when the input becomes empty.
- When we type "* " in the paragraph it is transformed into a list. If we press backspace the list should be transformed back to a paragraph. On this branch it does not happen anymore. The issue is real I verified not only using the end to end tests but also by manual testing. This seems totally unexpected and unrelated to the changes I'm not sure of what may be causing the issue.

saramarcondes commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -53,17 +53,15 @@ function getSelectOptions( optionsArray, disableCustomFontSizes ) {
+	value,
+	withSlider = false,
+	ref,
+} ) {
```
Why this change to move `ref` into the props themselves? I thought forwarded refs were always passed as a second parameter to the functional component?

jorgefilipecosta replied:
Thank you @saramarcondes, it was fixed 👍

jorgefilipecosta commented:
I added fixes to the bugs I identified. I needed to skip an end to end tests temporarily until the g2 fix I submitted is merged https://github.com/ItsJonQ/g2/pull/204.

I included the fix https://github.com/ItsJonQ/g2/pull/201 directly in this PR.

jorgefilipecosta commented:
The remaining issue happens because we are including none ECMA 5 code in our build.

```
/home/runner/work/gutenberg/gutenberg/build/components/index.js
6
  1:164942  error  Parsing error: The keyword 'const' is reserved
7
8
✖ 1 problem (1 error, 0 warnings)
```

The only non-compliment code is in array-move that is used by use-enhanced-state a dependency of g2. I'm not sure how we can force the transpiling of this code.

ItsJonQ commented:
> The only non-compliment code is in array-move that is used by use-enhanced-state a dependency of g2. 

@jorgefilipecosta I just tested this with the release of [0.0.125](https://github.com/ItsJonQ/g2/releases/tag/v0.0.125). `array-move` has been removed as a dependency.

The issue still persists. I believe it's coming from `zustand`, originally [mentioned here](https://github.com/WordPress/gutenberg/pull/27594#issuecomment-742652676).

@gziolo [mentioned](https://github.com/WordPress/gutenberg/pull/27594#issuecomment-742658089):

> I think webpack defaults to module field rather than main with our setup. You probably can add an exception to include this package in Babel transpilation.

I'm not sure how to do that though

jorgefilipecosta commented:
> The issue still persists. I believe it's coming from zustand, originally mentioned here.


Yes, we had an issue with array-move too but now it is fixed. Now it seems we still have the problem with zustand.


I changed our webpack config file packages/scripts/config/webpack.config.js to not exclude node_modules, using a function that excluded everything besides a module that needs transpiling, removing the exclude, switching the exclude to another folder besides node_modules, using another rule that included the modules we need to transpile, and nothing worked.
I'm not sure how we can force the transpiling of zustand, cc: @gziolo, @jsnajdr, @saramarcondes. Any help fixing this is appreciated :)

gziolo commented:
> I changed our webpack config file packages/scripts/config/webpack.config.js to not exclude node_modules, using a function that excluded everything besides a module that needs transpiling, removing the exclude, switching the exclude to another folder besides node_modules, using another rule that included the modules we need to transpile, and nothing worked.

Shouldn’t you apply changes in https://github.com/WordPress/gutenberg/blob/master/webpack.config.js? We shared some parts of the config for some time but it was too difficult to maintain so I think we only share code through plugins today.

jsnajdr commented:
After @jorgefilipecosta added the `zustand` transpilation, webpack started warning about missing export in `zustand/shallow`:
```
WARNING in ./packages/components/node_modules/@wp-g2/substate/dist/es/index.js 2:0-60
“export ‘default’ (reexported as ‘shallowCompare’) was not found in ‘zustand/shallow’
```

That's because `zustand/shallow` is a CommonJS module with some ESM interop definitions:
```js
Object.defineProperty(exports, '__esModule', { value: true });
function shallow(objA, objB) {
  if (typeof objA === 'object' && ... ) { ... }
}
exports.default = shallow;
```
Babel transpilation adds an `import` with a helper:
```js
import _typeof from "@babel/runtime/helpers/typeof";
Object.defineProperty(exports, '__esModule', { value: true });
function shallow(objA, objB) {
  if (_typeof(objA) === 'object' && ... ) { ... }
}
exports.default = shallow;
```
This makes the module a mix of ESM and CJS import/export syntax and webpack doesn't like that.

I've seen this before in Calypso and we'll need some Babel config tweaks to fix it. I'll try to remember what the fix is 🙂

youknowriad commented:
The change 38d548d is actually way more impactful than it looks.

It basically means `@wordpress/components` is not IE11 ready (like all of our packages) and that we're forcing people to drop support of IE11 or have a similar config to ours (add an exception for zustand). I believe this is the first time we break this consistency in our packages. 

Curious what you all think about that.

jsnajdr commented:
> It basically means `@wordpress/components` is not IE11 ready

Yes, if the IE11-ready promise we are making is not just that the `@wordpress/components` package code is IE11-compatible, but also that all its dependencies (even transitive ones) are carefully vetted to be IE11-compatible, too, then we're now breaking it.

jsnajdr commented:
https://github.com/WordPress/gutenberg/pull/27594/commits/0db3696d26f9de31979c469823caf98613ae6542 fixes the `zustand` transpilation. We need to use Babel's `sourceType` option to tell it whether it should treat the modules as ESM, CJS, or autodetect that.

Babel docs for `sourceType` describe exactly our case where adding `import` statements to a CJS module can confuse webpack: https://babeljs.io/docs/en/options#sourcetype

jsnajdr commented:
@ItsJonQ By the way, is `zustand` really something we can't live without? I didn't look at the library in much detail, but after a quick look, it sounds plausible that it could be replaced with a native `React.useReducer()`.

gziolo commented:
This is all the code that `zustand` ships:

```js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _regeneratorRuntime = require('@babel/runtime/regenerator');
var react = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);

function create(createState) {
  var state;
  var listeners = new Set();

  var setState = function setState(partial, replace) {
    var nextState = typeof partial === 'function' ? partial(state) : partial;

    if (nextState !== state) {
      var previousState = state;
      state = replace ? nextState : Object.assign({}, state, nextState);
      listeners.forEach(function (listener) {
        return listener(state, previousState);
      });
    }
  };

  var getState = function getState() {
    return state;
  };

  var subscribeWithSelector = function subscribeWithSelector(listener, selector, equalityFn) {
    if (selector === void 0) {
      selector = getState;
    }

    if (equalityFn === void 0) {
      equalityFn = Object.is;
    }

    var currentSlice = selector(state);

    function listenerToAdd() {
      var nextSlice = selector(state);

      if (!equalityFn(currentSlice, nextSlice)) {
        var previousSlice = currentSlice;
        listener(currentSlice = nextSlice, previousSlice);
      }
    }

    listeners.add(listenerToAdd); // Unsubscribe

    return function () {
      return listeners.delete(listenerToAdd);
    };
  };

  var subscribe = function subscribe(listener, selector, equalityFn) {
    if (selector || equalityFn) {
      return subscribeWithSelector(listener, selector, equalityFn);
    }

    listeners.add(listener); // Unsubscribe

    return function () {
      return listeners.delete(listener);
    };
  };

  var destroy = function destroy() {
    return listeners.clear();
  };

  var api = {
    setState: setState,
    getState: getState,
    subscribe: subscribe,
    destroy: destroy
  };
  state = createState(setState, getState, api);
  return api;
}

var useIsoLayoutEffect = typeof window === 'undefined' ? react.useEffect : react.useLayoutEffect;
function create$1(createState) {
  var api = typeof createState === 'function' ? create(createState) : createState;

  var useStore = function useStore(selector, equalityFn) {
    if (selector === void 0) {
      selector = api.getState;
    }

    if (equalityFn === void 0) {
      equalityFn = Object.is;
    }

    var _useReducer = react.useReducer(function (c) {
      return c + 1;
    }, 0),
        forceUpdate = _useReducer[1];

    var state = api.getState();
    var stateRef = react.useRef(state);
    var selectorRef = react.useRef(selector);
    var equalityFnRef = react.useRef(equalityFn);
    var erroredRef = react.useRef(false);
    var currentSliceRef = react.useRef();

    if (currentSliceRef.current === undefined) {
      currentSliceRef.current = selector(state);
    }

    var newStateSlice;
    var hasNewStateSlice = false; // The selector or equalityFn need to be called during the render phase if
    // they change. We also want legitimate errors to be visible so we re-run
    // them if they errored in the subscriber.

    if (stateRef.current !== state || selectorRef.current !== selector || equalityFnRef.current !== equalityFn || erroredRef.current) {
      // Using local variables to avoid mutations in the render phase.
      newStateSlice = selector(state);
      hasNewStateSlice = !equalityFn(currentSliceRef.current, newStateSlice);
    } // Syncing changes in useEffect.


    useIsoLayoutEffect(function () {
      if (hasNewStateSlice) {
        currentSliceRef.current = newStateSlice;
      }

      stateRef.current = state;
      selectorRef.current = selector;
      equalityFnRef.current = equalityFn;
      erroredRef.current = false;
    });
    var stateBeforeSubscriptionRef = react.useRef(state);
    react.useEffect(function () {
      var listener = function listener() {
        try {
          var nextState = api.getState();
          var nextStateSlice = selectorRef.current(nextState);

          if (!equalityFnRef.current(currentSliceRef.current, nextStateSlice)) {
            stateRef.current = nextState;
            currentSliceRef.current = nextStateSlice;
            forceUpdate();
          }
        } catch (error) {
          erroredRef.current = true;
          forceUpdate();
        }
      };

      var unsubscribe = api.subscribe(listener);

      if (api.getState() !== stateBeforeSubscriptionRef.current) {
        listener(); // state has changed before subscription
      }

      return unsubscribe;
    }, []);
    return hasNewStateSlice ? newStateSlice : currentSliceRef.current;
  };

  Object.assign(useStore, api); // For backward compatibility (No TS types for this)

  useStore[Symbol.iterator] = /*#__PURE__*/_regeneratorRuntime__default['default'].mark(function _callee() {
    return _regeneratorRuntime__default['default'].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            console.warn('[useStore, api] = create() is deprecated and will be removed in v4');
            _context.next = 3;
            return useStore;

          case 3:
            _context.next = 5;
            return api;

          case 5:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  });
  return useStore;
}

exports.default = create$1;
```

Do we use something else that builds on top of it?

ItsJonQ commented:
> @ItsJonQ By the way, is zustand really something we can't live without? I didn't look at the library in much detail, but after a quick look, it sounds plausible that it could be replaced with a native React.useReducer().

TLDR, we need it. We either need `zustand` or something that handles state/render cycles in a similar way.

The Components were originally constructed with `useState`/`useReducer`, but we ran into performance issues. Zustand has helped tremendously with performance and internal component state management.

I shared some of my performance investigations here:
https://github.com/ItsJonQ/g2/issues/57#issuecomment-706606337

The performance issues were resolved by leveraging a listener/subscriber flow, helping with the React render cycles. cc'ing @diegohaz as I think he may have some more insight into this :).

@jsnajdr I believe state management is something you have a lot of experience in (🤗 ). Hopefully the strategy I mentioned above helps provide context in some way.

---

For compiling reasons, if we can't (easily) consume `zustand` from `node_modules`, then I would personally fork that library and use it that way.

gziolo commented:
It looks like `@wordpres/stan` from @youknowriad could be a good fit here then 😄

diegohaz commented:
Just noting that I've built this tiny implementation of a pubsub state with React (in a way that works with concurrent mode): https://codesandbox.io/s/react-pubsub-state-gpc1u?file=/src/App.js

But I think leveraging `@wordpress/stan` could be a good idea too.

youknowriad commented:
Something as minimal as possible makes sense to me, I'd have preferred just local state but I don't know enough about the complexity of the usage of this in the components to judge. Not sure the `@wordpress/components` package should use any elaborate data package.

ItsJonQ commented:
> Something as minimal as possible makes sense to me, I'd have preferred just local state but I don't know enough about the complexity of the usage of this in the components to judge. Not sure the @wordpress/components package should use any elaborate data package.

For context, how `zustand` is used in `@wp-g2` is local state.
It can thought of as being an enhanced `useReducer`, with improved performance and portability (within sub-components).

At this point, what direction would we like to go down?

If there's a concern regarding babel and IE11 (which is valid), then I can copy/paste the `zustand` library into `@wp-g2/substate` rather than using it as a dependency.

If using an alternative state management solution is a concern, then we'll have to refactor a lot of things within `@wp-g2/components` to use `React.useState` and `React.useReducer` (which were the original implementations before running into some performance issues).

jsnajdr commented:
> The Components were originally constructed with useState/useReducer, but we ran into performance issues.

I understand now why `useReducer` would be significantly slower than `zustand`: any state update of the reducer is propagated to all its consumer components, and triggers their rerender. On the other hand, `zustand` supports selector that select only a slice of the state and trigger a React rerender only when that slice changes. That saves a lot of wasted rerenders.

However, that applies only when one store is shared by multiple components using a React context. That's what the [G2 performance test](https://github.com/ItsJonQ/g2/tree/master/packages/design-tools/src/__stories__/PerformanceTest) does and I can see why it gets much slower with vanilla `useReducer`.

Usages of `useSubState()` in the G2 components work very differently though. They manage only one component's local state, the store is never seen by any other component and the selector optimization doesn't save any rerenders. Plain `useState` or `useReducer` should work here equally well without any performance penalty. Hooks like `useAccordion`, `useColorPickerState` or `useTextInputState` could be implemented like this, without a 3rd party dependency.

Then there are usages with `createStore` and context. Like `useModalStore`, `useThemeStylesStore` or `useContextSystem`. These could theoretically make use of the selector optimization, but in practice they don't. These stores store only one atomic value (`theme`, `context`) and consumers use it in such a way that the rerender is triggered on any change. That's suboptimal particularly for the context system. `useContextSystem( props, 'Alert' )` and `useContextSystem( props, 'Accordion' )`, should never trigger cross-namespace updates, but they do.

So, the conclusion so far is that the performance test is testing something that G2 doesn't really do in practice and that `useReducer` should be equally fast in most cases.

Still, the selector optimization is valuable. It would certainly make `useContextSystem` faster, at least.

On the other hand, Redux and `@wordpress/data` support this optimization, too, and I don't see a fundamental reason why any of them should do it faster than the other. They all implement more or less the same thing.

For these usages, I think that a `@wordpress/data` store would be a suitable and well-performing replacement for `zustand`.

The anonymous unregistered data stores that @gziolo started working on in #26655 could also be useful here. You don't need to register a global singleton store like `core/components-context`. You can just create a data store, assign the return value ("store descriptor") to a variable, pass that variable around however you want to (module exports, context, ...), and use it to work with the store using the `useSelect( storeDesc )` and `useDispatch( storeDesc )` hook.

jsnajdr commented:
> If there's a concern regarding babel and IE11 (which is valid), then I can copy/paste the zustand library into `@wp-g2/substate` rather than using it as a dependency.

That will let us avoid transpiling `node_modules` dependencies, at least for now. But I'm afraid that one day we'll need to start doing it anyway, and there's no point trying to delay this addition. Published NPM modules with ES6+ syntax are increasingly common. And it's hard to argue that there's something wrong with these modules and that they should be fixed. Why would ES5 syntax be "right" and ES6 syntax, with `let` and `const`, be "wrong"? It's completely arbitrary. The ES6 standard is 5 years old, and Node.js has been supporting it natively for a long time.

youknowriad commented:
@jsnajdr I don't think it's wrong to ship ES6 by default. I think it's different from what we're doing now and not a small change. We're changing the expectations of consumers.

 - It's a breaking change of course 
 - It seems that it's something that should be handled in `wp-scripts build` in case third-party packages are relying on it to ship code  to IE11 (which is what the current browser list of wp-scripts says)

The other thing is alignment with the JS community as a whole, what's the current expectations for npm users and on that aspect it seems like we're in the beginning of some kind of transition period. Most packages still ship ES5 code but some are starting to ship ES6. 

And when it comes to bundlers, most of them are still not transpiling node_modules by default (webpack, parcel...) but some of them are starting to changing the default here (CRA for instance).

jsnajdr commented:
> The other thing is alignment with the JS community as a whole, what's the current expectations for npm users and on that aspect it seems like we're in the beginning of some kind of transition period.

Regarding the alignment with community, it seems to me the community is in a kind of a deadlock state where everyone is waiting for everyone else to make the first move. And that's been going on for years. Babel maintainers have published a [blog post about it](https://babeljs.io/blog/2018/06/26/on-consuming-and-publishing-es2015+-packages) in June 2018.

The result is that, while searching for a perfect solution and rejecting imperfect ones, the vast majority of users, with evergreen browsers and Node runtimes, are consuming suboptimal code with a lot of unneeded polyfills (e.g., regenerator runtime).

jsnajdr commented:
> I can copy/paste the `zustand` library into `@wp-g2/substate` rather than using it as a dependency.

This looks like the fastest solution that makes everyone happy. In a longer term, we should consider reusing `@wordpress/data` in the G2 components, and/or just work with local state and `useReducer`. And also solve the issue with transpiling `node_modules`. Neither of these is a quick fix though.

ItsJonQ commented:
@jsnajdr I really appreciate you digging into the G2 Components codebase so deeply 🙏 .

Your suggestions have given me some more insight into how I could revisit state management in the various systems, especially the context one. I'm not sure what the immediate solutions are. 

I'm hopeful we'll be able to achieve a solution or convention that harmonizes performance, portability, and ease of use (and understanding).

Thanks @gziolo , @youknowriad , @diegohaz , @jorgefilipecosta , and @saramarcondes for looking into this ❤️

ItsJonQ commented:
It looks like the tests are stuck on the generator function from `zustand`:
https://github.com/ItsJonQ/g2/blob/master/packages/substate/src/zustand/index.js#L121

I believe I'll need to transform this over in the `@wp-g2` repo to make it safe for Gutenberg to consume.

What would the best babel plugin (or convention) to do this :)

jorgefilipecosta commented:
After the fork of zustand it seems we don't have transpiling issues anymore.

But unfortunately, our unit tests are now failing and it seems related to zustand:
```
ReferenceError: regeneratorRuntime is not defined

      2 |  * External dependencies
      3 |  */
    > 4 | import { ContextSystemProvider } from '@wp-g2/components';
        | ^
      5 | 
      6 | export function ComponentSystemProvider( {
      7 | 	__unstableNextInclude = [],

      at create (node_modules/@wp-g2/substate/dist/cjs/zustand/index.js:160:44)
      at createContextSystemStore (node_modules/@wp-g2/context/dist/cjs/context-system-provider.js:43:54)
      at Object.<anonymous> (node_modules/@wp-g2/context/dist/cjs/context-system-provider.js:62:30)
      at Object.<anonymous> (node_modules/@wp-g2/context/dist/cjs/index.js:7:30)
      at Object.<anonymous> (node_modules/@wp-g2/components/dist/cjs/ContextSystem/index.js:7:16)
      at Object.<anonymous> (node_modules/@wp-g2/components/dist/cjs/index.js:7:22)
      at Object.<anonymous> (packages/components/src/__next/context/component-system-provider.js:4:1)
      at Object.<anonymous> (packages/components/src/__next/context/index.js:1:1)
      at Object.<anonymous> (packages/components/src/font-size-picker/next/index.js:4:1)
      at Object.<anonymous> (packages/components/src/font-size-picker/index.js:21:1)
      at Object.<anonymous> (packages/components/src/index.js:68:1)
      at Object.<anonymous> (packages/block-editor/src/components/inserter/index.js:12:1)
      at Object.<anonymous> (packages/block-editor/src/components/default-block-appender/index.js:18:1)
      at Object.<anonymous> (packages/block-editor/src/components/default-block-appender/test/index.js:9:1)
```


@ItsJonQ, @saramarcondes any idea of what may be causing this issue?

saramarcondes commented:
@jorgefilipecosta I believe that's what @ItsJonQ was talking about in the comment immediately previous to your's.

jorgefilipecosta commented:
Ups, I'm sorry, I had the tab open for a long time, and for some reason, it did auto-updated to include the comment.

ItsJonQ commented:
Updating the build of `@wp-g2/` to account for babel runtime 🤞 
https://github.com/ItsJonQ/g2/pull/228

Commit:
Update @wp-g2 packages to latest 0.0.135 with babel runtime handling

ItsJonQ commented:
I just updated the `@wp-g2` packages to [v0.0.135](https://github.com/ItsJonQ/g2/releases/tag/v0.0.135) which is now built with babel runtime (to resolve the regenerator issue).

Things seem to be working okay from my tests. In addition to Gutenberg,I've tested this built with multiple projects in different environments (CodeSandbox, local build, and other cloud/CI platforms)

🤞

ItsJonQ commented:
The Static Analysis test is failing at:

```
npm run check-local-changes
```


```
There are local uncommitted changes after one or both of 'npm install' or 'npm run docs:build'!
diff --git a/package-lock.json b/package-lock.json
```

Not too sure how to resolve this one 🤔 .
On the plus side, the unit test is passing!

gziolo commented:
`npm run docs:build` – you run it and commit changes if there is any `*.md` file updated.

When `npm install` is the reason, it usually helps to `rm -rf node_modules` and then `npm i` with the latest `npm` version.

Commit:
Update package-lock.json

Commit:
Update @wp-g2 to latest 0.0.139

ItsJonQ commented:
I just pushed an update with the `@wp-g2` packages at [0.0.139](https://github.com/ItsJonQ/g2/releases/tag/v0.0.139). With this version, `zustand` has been completely refactored out. Also, the `@wordpress/*` dependencies have been correctly set as `peerDependencies`

ItsJonQ commented:
OMG.... are the tests... all 🍏 GREEN 🍏 ?!

gziolo commented on packages/e2e-tests/specs/editor/various/font-size-picker.test.js:
```diff
@@ -98,7 +98,7 @@ describe( 'Font Size Picker', () => {
 	} );
 
-	it( 'should reset a named font size using input field', async () => {
+	it.skip( 'should reset a named font size using input field', async () => {
```
What's the status of this test. Should we enable it, update or remove it completely?

ItsJonQ replied:
Oh, hmm! I didn't know about this. Let me try re-enabling

Commit:
Re-enabled E2E font-size-picker reset test

ItsJonQ commented:
Oh boy! All green all around! Thanks for approving @saramarcondes !

@gziolo Looks like we may be ready to land this sometime this week :)

gziolo reviewed (approved):
From my perspective you can merge it even today 😃

ItsJonQ commented:
> From my perspective you can merge it even today 😃

In that case... I'm gonna do it! 🚀 🤞

Commit:
Update @wp-g2 to v0.0.140 - Removing @wp-g2/substate as a dependency

Commit:
Refactor usage of noop and is utils

gziolo commented:
Exciting 🎉✨🔥

youknowriad commented:
Thanks all. Nice work. We're just getting started

gziolo commented:
I was looking again at the code merged to idenfity all the G2 components used that we should work on porting next:

- `Button` (imported 1 time)
- `ControlLabel` (1)
- `Grid` (2)
- `FormGroup` (1)
- `SelectDropdown` (1)
- `Slider` (1)
- `TextInput` (2)
- `View` (2)
- `VisuallyHidden` (1)
- `VStack` (2)

@ItsJonQ, what would be your recommendation to tackle next based on the complexity. The lower the better I believe.

There are also two imports from `@wp-gw/utils` that we could replace with either lodash or native JavaScript code:
- `is.defined` (by the way, grouped imports prevent tree-shaking)
-  `noop`

We should also inline two another utils inside `@wordpress/components` for now:
- `createUnitValue`
- `parseUnitValue`

ItsJonQ commented:
@gziolo Thank you for the follow up!

>  tackle next based on the complexity. The lower the better I believe. 

I agree! I think we should port these components one at a time.

I created a new issue with details on how we can continue with this migration effort:
https://github.com/WordPress/gutenberg/issues/28399

In it, I've listed the components we'll need to migrate in order of dependency and complexity.

> There are also two imports from @wp-gw/utils that we could replace with either lodash or native JavaScript code:

Ah yes. We'll need to refactor the component code when we move things order to use the conventions of this codebase.

> We should also inline two another utils inside @wordpress/components for now

That seems like something we can do when we move over the components that have unit parsing in them.

gziolo commented:
Tracking issue looks great, thanks 🙇‍♂️
